File: Python/Trees.py
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def dfs_inorder(root):
    if not root:
        return
    dfs_inorder(root.left)
    print(root.val, end=" ")
    dfs_inorder(root.right)

def dfs_preorder(root):
    if not root:
        return
    print(root.val, end=" ")
    dfs_preorder(root.left)
    dfs_preorder(root.right)

def dfs_postorder(root):
    if not root:
        return
    dfs_postorder(root.left)
    dfs_postorder(root.right)
    print(root.val, end=" ")

def bfs(root):
    if not root:
        return
    queue = [root]
    while queue:
        node = queue.pop(0)
        print(node.val, end=" ")
        if node.left:
            queue.append(node.left)
        if node.right:
            queue.append(node.right)

# Create a binary tree
root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)
root.left.left = TreeNode(4)
root.left.right = TreeNode(5)
root.right.left = TreeNode(6)
root.right.right = TreeNode(7)

# Binary Tree Traversals
print("DFS Inorder:", end=" ")
dfs_inorder(root)
print()

print("DFS Preorder:", end=" ")
dfs_preorder(root)
print()

print("DFS Postorder:", end=" ")
dfs_postorder(root)
print()

print("BFS:", end=" ")
bfs(root)
print()

Binary Tree Traversals:

DFS Inorder:

Time Complexity: O(N) - where N is the number of nodes in the binary tree.
Space Complexity: O(H) - where H is the height of the binary tree.
In the worst case, the space complexity can be O(N) for a skewed binary tree.

DFS Preorder:

Time Complexity: O(N) - where N is the number of nodes in the binary tree.
Space Complexity: O(H) - where H is the height of the binary tree.
In the worst case, the space complexity can be O(N) for a skewed binary tree.

DFS Postorder:

Time Complexity: O(N) - where N is the number of nodes in the binary tree.
Space Complexity: O(H) - where H is the height of the binary tree.
In the worst case, the space complexity can be O(N) for a skewed binary tree.

BFS:

Time Complexity: O(N) - where N is the number of nodes in the binary tree.
Space Complexity: O(W) - where W is the maximum width of the binary tree at any level.

File: C++/Trees.cpp
#include <iostream>
#include <queue>

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode(int val) : val(val), left(nullptr), right(nullptr) {}
};

// Depth-First Search (DFS) Inorder Traversal
void dfsInorder(TreeNode* root) {
    if (!root) return;
    dfsInorder(root->left);
    std::cout << root->val << " ";
    dfsInorder(root->right);
}

// Depth-First Search (DFS) Preorder Traversal
void dfsPreorder(TreeNode* root) {
    if (!root) return;
    std::cout << root->val << " ";
    dfsPreorder(root->left);
    dfsPreorder(root->right);
}

// Depth-First Search (DFS) Postorder Traversal
void dfsPostorder(TreeNode* root) {
    if (!root) return;
    dfsPostorder(root->left);
    dfsPostorder(root->right);
    std::cout << root->val << " ";
}

// Breadth-First Search (BFS) Traversal
void bfs(TreeNode* root) {
    if (!root) return;
    std::queue<TreeNode*> q;
    q.push(root);
    while (!q.empty()) {
        TreeNode* curr = q.front();
        q.pop();
        std::cout << curr->val << " ";
        if (curr->left) q.push(curr->left);
        if (curr->right) q.push(curr->right);
    }
}

int main() {
    // Create a binary tree
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);
    root->right->left = new TreeNode(6);
    root->right->right = new TreeNode(7);

    // Binary Tree Traversals
    std::cout << "DFS Inorder: ";
    dfsInorder(root);
    std::cout << std::endl;

    std::cout << "DFS Preorder: ";
    dfsPreorder(root);
    std::cout << std::endl;

    std::cout << "DFS Postorder: ";
    dfsPostorder(root);
    std::cout << std::endl;

    std::cout << "BFS: ";
    bfs(root);
    std::cout << std::endl;

    return 0;
}

Using shared_ptr:

#include <iostream>
#include <memory>
#include <queue>

struct TreeNode {
    int val;
    std::shared_ptr<TreeNode> left;
    std::shared_ptr<TreeNode> right;

    TreeNode(int val) : val(val), left(nullptr), right(nullptr) {}
};

// Depth-First Search (DFS) Inorder Traversal
void dfsInorder(const std::shared_ptr<TreeNode>& root) {
    if (!root) return;
    dfsInorder(root->left);
    std::cout << root->val << " ";
    dfsInorder(root->right);
}

// Depth-First Search (DFS) Preorder Traversal
void dfsPreorder(const std::shared_ptr<TreeNode>& root) {
    if (!root) return;
    std::cout << root->val << " ";
    dfsPreorder(root->left);
    dfsPreorder(root->right);
}

// Depth-First Search (DFS) Postorder Traversal
void dfsPostorder(const std::shared_ptr<TreeNode>& root) {
    if (!root) return;
    dfsPostorder(root->left);
    dfsPostorder(root->right);
    std::cout << root->val << " ";
}

// Breadth-First Search (BFS) Traversal
void bfs(const std::shared_ptr<TreeNode>& root) {
    if (!root) return;
    std::queue<std::shared_ptr<TreeNode>> q;
    q.push(root);
    while (!q.empty()) {
        std::shared_ptr<TreeNode> curr = q.front();
        q.pop();
        std::cout << curr->val << " ";
        if (curr->left) q.push(curr->left);
        if (curr->right) q.push(curr->right);
    }
}

int main() {
    // Create a binary tree
    auto root = std::make_shared<TreeNode>(1);
    root->left = std::make_shared<TreeNode>(2);
    root->right = std::make_shared<TreeNode>(3);
    root->left->left = std::make_shared<TreeNode>(4);
    root->left->right = std::make_shared<TreeNode>(5);
    root->right->left = std::make_shared<TreeNode>(6);
    root->right->right = std::make_shared<TreeNode>(7);

    // Binary Tree Traversals
    std::cout << "DFS Inorder: ";
    dfsInorder(root);
    std::cout << std::endl;

    std::cout << "DFS Preorder: ";
    dfsPreorder(root);
    std::cout << std::endl;

    std::cout << "DFS Postorder: ";
    dfsPostorder(root);
    std::cout << std::endl;

    std::cout << "BFS: ";
    bfs(root);
    std::cout << std::endl;

    return 0;
}
